### PR TITLE
Occupant standby mode control

### PR DESCRIPTION
**Pull request by lymereJ**
The proposed changes apply standby mode controls to space types targeted by ASHRAE 90.1-2019. This implementation uses a simplified approach where controls are only modeled for one hour each day, the occupancy profile for the targeted spaces is left unchanged. The standby mode control require a setup/setback of the thermostat setpoint and shutoff of the ventilation. The shutoff of the ventilation for zones served by unitary systems is done by setting the fan operation to cycle during that hour and for zone served by a VAV box, the minimum damper position is scheduled to a low value during that hour. Additional information regarding about this requirement are provided in Section B2.2 in the Energy Savings Analysis: [ANSI/ASHRAE/IES Standard 90.1-2019 report](https://www.energycodes.gov/sites/default/files/2021-07/20210407_Standard_90.1-2019_Determination_TSD.pdf).

### CODE REVIEW

File: lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2016/ashrae_90_1_2016.PumpVariableSpeed.rb [hunk 1]
```diff
@@ -76,7 +76,7 @@ def pump_variable_speed_get_control_type(pump, plant_loop_type, pump_nominal_hp)
                'ASHRAE 169-2013-1A'
             threshold = 150
           else
-            OpenStudio.logFree(OpenStudio::Warn, "Pump flow control requirement missing for heating water pumps in climate zone: #{climate_zone}.")
+            OpenStudio.logFree(OpenStudio::Warn, 'openstudio.ashrae_90_1_2016.PumpVariableSpeed', "Pump flow control requirement missing for heating water pumps in climate zone: #{climate_zone}.")
```
Comment (lymereJ) [inline lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2016/ashrae_90_1_2016.PumpVariableSpeed.rb:79]: Unrelated but necessary logging changes.

File: lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    # Currently standards doesn't excatly use the 90.1
+    # space description so all spaces types that include
+    # office/meeting/conference are flagged as having
+    # occupant standby mode.
+    if std_space_type.downcase.include?('office') || std_space_type.downcase.include?('meeting') || std_space_type.downcase.include?('conference')
+      return true
+    end
+
+    return false
+  end
```
Comment (lymereJ) [inline lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb:34]: New function that identifies spaces required to have occupant standby controls.

File: lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    if clg_sch_old.empty?
+      clg_sch_offset = model_offset_schedule_value(clg_sch, clg_sch_mod)
+      clg_sch_offset.setName(clg_sch_name)
+      thermostat.setCoolingSchedule(clg_sch_offset)
+    else
+      thermostat.setCoolingSchedule(clg_sch_old.get)
+    end
+
+    return true
+  end
```
Comment (lymereJ) [inline lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb:75]: New function that adjust the thermostat of zone with spaces required to have occupant standby mode controls.

File: lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb [hunk 4]
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
...
+          day_schedule.addValue(t, value)
+          t_p_1 = OpenStudio::Time.new(0, hr + 1, 0, 0)
+          day_schedule.addValue(t_p_1, value + time_offset_hash[hr.to_s])
+        end
+      end
+    end
+
+    offset_sch.setName("#{schedule.name} - offset")
+    return offset_sch
+  end
```
Comment (lymereJ) [inline lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:2599]: New method to offset schedule values using a hash that associate the time of the day to the offset value.

File: lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb [hunk 5]
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
...
+          day_schedule.addValue(t, value)
+          t_p_1 = OpenStudio::Time.new(0, hr + 1, 0, 0)
+          day_schedule.addValue(t_p_1, time_value_hash[hr.to_s])
+        end
+      end
+    end
+
+    new_sch.setName("#{schedule.name} - adjusted")
+    return new_sch
+  end
```
Comment (lymereJ) [inline lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:2641]: New method that sets a values of a schedule. Time are value are provided into a hash.

File: lib/openstudio-standards/standards/Standards.ZoneHVACComponent.rb [hunk 6]
```diff
@@ -171,6 +171,35 @@ def zone_hvac_component_apply_standard_controls(zone_hvac_component)
...
+    standby_mode_spaces = []
+    thermal_zone.spaces.sort.each do |space|
+      if space_is_occupancy_standby_mode(space)
+        standby_mode_spaces << space
+      end
+    end
+    if !standby_mode_spaces.empty?
+      zone_hvac_model_standby_mode_occupancy_control(zone_hvac_component)
+    end
+
```
Comment (lymereJ) [inline lib/openstudio-standards/standards/Standards.ZoneHVACComponent.rb:202]: Handles occupant standby control for zonal systems.

File: lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.AirLoopHVAC.rb [hunk 7]
```diff
@@ -453,4 +453,67 @@ def air_loop_hvac_energy_recovery_ventilator_flow_limit(air_loop_hvac, climate_z
...
+            end
+          else
+            OpenStudio.logFree(OpenStudio::Warn, 'openstudio.ashrae_90_1_2019.AirLoopHVAC', "The air terminal associated with #{zone.name} isn't of the SingleDuctVAVReheat type so occupant standby controls were not modeled.")
+          end
+        end
+      end
+    end
+
+    return true
+  end
```
Comment (lymereJ) [inline lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.AirLoopHVAC.rb:518]: New method that sets fan operating mode of unitary system to be cycling during occupant standby mode period, and schedule MDP of VAV boxes to a lower value during occupant standby mode period.

**Review by mdahlhausen — CHANGES_REQUESTED** [review]
space_occupancy_standby_mode(thermostat) is similar in naming to space_is_occupancy_standby_mode(space).
We've generally used _required? at the end of methods to check if a certain control scheme is required.  Ample examples in AirLoopHVAC.rb.

Also, the documentation is confusing in [Prototype.Model.rb](lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb):
# Modify thermostat schedule to account for a thermostat setback/up
#
# @param space [OpenStudio::model::Space] OpenStudio Space object
# @return [Boolean] true if success
def space_occupancy_standby_mode(thermostat)
  return true
end

@param space should be @param thermostat.  And why does it default to true instead of false?

Some other thoughts - the setback time is arbitrary, but could be knowable in the model from the schedule hours of operation additional properties object, or from the methods that generate a boolean zone occupancy.  Note sure how to address that.  But I suspect this may cause some issues with more generic models applying the 90.1-2019 logic.

**Comment by lymereJ** [comment]
> space_occupancy_standby_mode(thermostat) is similar in naming to space_is_occupancy_standby_mode(space).
> We've generally used _required? at the end of methods to check if a certain control scheme is required. Ample examples in AirLoopHVAC.rb.

That makes sense. Addressed in the latest commit.

> Also, the documentation is confusing in [Prototype.Model.rb](lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb):
> 
> ```
> # Modify thermostat schedule to account for a thermostat setback/up
> #
> # @param space [OpenStudio::model::Space] OpenStudio Space object
> # @return [Boolean] true if success
> def space_occupancy_standby_mode(thermostat)
>   return true
> end
> ```
> 
> @param space should be @param thermostat. And why does it default to true instead of false?

Good catch, I changed my approach and forgot to change the docs.

> Some other thoughts - the setback time is arbitrary, but could be knowable in the model from the schedule hours of operation additional properties object, or from the methods that generate a boolean zone occupancy. Note sure how to address that. But I suspect this may cause some issues with more generic models applying the 90.1-2019 logic.

I agree, do you think that it would make sense to open an issue and keep this PR as-is for now?

**Review by mdahlhausen — CHANGES_REQUESTED** [review]
Changes look good.

I think these methods should be generalized to take a zone or space occupancy schedule to create the offset:
`model_offset_schedule_value(schedule, time_offset_hash)`
`space_occupancy_standby_mode(thermostat)`

We can hold off on the merge for now.

**Comment by lymereJ** [comment]
@mdahlhausen  - `model_offset_schedule_value(schedule, time_offset_hash)` was meant to be a helper function that could be used by any other function. Refactoring to take a zone of space occupancy schedule seems like it would reduce its scope of applicability. Don't you think?

If we want to generalize `space_occupancy_standby_mode(thermostat)` we also need to generalize `zone_hvac_model_standby_mode_occupancy_control(zone_hvac_component)`. When you say hold off on the merge, do you mean as part of the next release?

File: lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  # @return [Boolean] true if success
+  def space_occupancy_standby_mode(thermostat)
+    htg_sch = thermostat.getHeatingSchedule.get
+    clg_sch = thermostat.getCoolingSchedule.get
+
+    # Setback heating schedule
+    # Setback is 1 deg. F per code requirement
+    # Time of the day is arbitrary lack of dynamic occupant modeling
+    setup = 1 # deg. F
+    htg_sch_mod = { '12' => -1 * (OpenStudio.convert(1, 'F', 'C').get - OpenStudio.convert(1 - setup, 'F', 'C').get) }
```
Comment (mdahlhausen) [inline lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb:48]: use Ranking and Kelvin, not Farenheit and Celsius

File: lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb [hunk 9]
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
...
+        t = OpenStudio::Time.new(0, hr, 0, 0)
+
+        # Get schedule value
+        value = day_schedule.getValue(t)
+
+        # Offset schedule value
+        if time_offset_hash.key?(hr.to_s)
+          day_schedule.addValue(t, value)
+          t_p_1 = OpenStudio::Time.new(0, hr + 1, 0, 0)
+          day_schedule.addValue(t_p_1, value + time_offset_hash[hr.to_s])
```
Comment (mdahlhausen) [inline lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:2592]: does this override?

File: lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb [hunk 10]
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
...
+    default_day_schedule = offset_sch.defaultDaySchedule
+    day_schedules << default_day_schedule
+    offset_sch.scheduleRules.each do |rule|
+      day_schedules << rule.daySchedule
+    end
+
+    # Offset schedule values
+    day_schedules.each do |day_schedule|
+      (0..23).each do |hr|
+        t = OpenStudio::Time.new(0, hr, 0, 0)
```
Comment (mdahlhausen) [inline lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:2583]: guard clause here?

Comment (mdahlhausen) [reply lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:2583]: next if hr not in time_offset_hash

File: lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb (hunk 8 above)
Comment (lymereJ) [reply lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb]: Addressed in latest commit.

File: lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb (hunk 10 above)
Comment (lymereJ) [reply lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:2585]: Addressed in latest commit.

File: lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb (hunk 9 above)
Comment (lymereJ) [reply lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb]: No, because it sets it for the next hour.

**Review by mdahlhausen — APPROVED** [review]

**Pull request merged**